node/full_node_integration_test.go

```diff
@@ -379,16 +375,76 @@ func (s *FullNodeTestSuite) TestStateRecovery() {
 	// Wait for some blocks
 	time.Sleep(2 * s.node.nodeConfig.BlockTime)
 
-	// Restart node, we don't need to check for errors
-	_ = s.node.Stop(s.ctx)
-	_ = s.node.Start(s.ctx)
+	// Stop the current node
+	s.cancel()
+
+	// Wait for the node to stop
+	waitCh := make(chan struct{})
+	go func() {
+		s.runningWg.Wait()
+		close(waitCh)
+	}()
+
+	select {
+	case <-waitCh:
+		// Node stopped successfully
+	case <-time.After(2 * time.Second):
+		s.T().Log("Warning: Node did not stop gracefully within timeout")
+	}
+
+	// Create a new context
+	s.ctx, s.cancel = context.WithCancel(context.Background())
+	s.errCh = make(chan error, 1)
+
+	// Create a NEW node instance instead of reusing the old one
+	config := getTestConfig(1)
+	genesis, genesisValidatorKey := types.GetGenesisWithPrivkey(types.DefaultSigningKeyType, "test-chain")
+	signingKey, err := types.PrivKeyToSigningKey(genesisValidatorKey)
+	require.NoError(err)
+	p2pKey := generateSingleKey()
+
+	dummyExec := coreexecutor.NewDummyExecutor()
+	dummySequencer := coresequencer.NewDummySequencer()
+
+	node, err := NewNode(
+		s.ctx,
+		config,
+		dummyExec,
+		dummySequencer,
+		p2pKey,
+		signingKey,
+		genesis,
+		DefaultMetricsProvider(cmcfg.DefaultInstrumentationConfig()),
+		log.NewTestLogger(s.T()),
+	)
+	require.NoError(err)
+
+	fn, ok := node.(*FullNode)
+	require.True(ok)
+
+	// Replace the old node with the new one
+	s.node = fn
+
+	// Start the new node
+	s.runningWg.Add(1)
+	go func() {
+		defer s.runningWg.Done()
+		err := s.node.Run(s.ctx)
+		select {
+		case s.errCh <- err:
+		default:
+			s.T().Logf("Error channel full, discarding error: %v", err)
+		}
+	}()
```

Comment:
This (or very similar) code is repeated multiple times. Consider extracting small helper function.

Reply:
extracted,